### ISSUE TITLE
Proposal / Feature Request: Pin Repositories

### ISSUE THREAD

**Issue by dschier-wtd** [go-gitea/gitea#10375]
Hi team,
first let me say: It's awesome what you are doing here!

I wanted to propose a feature, currently existing in many other SCM UIs. Having the option to pi a specific amount of repositories to the top would be awesome.

## Description

Pinning a repository to the top of an organization can be helpful for many use cases.

- putting Skeletons / Templates / Conventions / Concepts on top
- featuring repositories in an organization
- highlighting important repositories

## Mockup / Ideas

I am seeing three possible options here to have a pinned repository:

### 1. Pinned Cards

This option will lead the viewers attention directly to the featured / pinned / important repositories.

- limitation will be needed  (3 - 8)
- organization of the cards should be possible

![cards](https://user-images.githubusercontent.com/6490538/74917879-864d5100-53c8-11ea-8080-f266c4ec1fc0.png)

### 2. Featured Repository

This option will take use of the unused space on the right side the same way as people and teams are featured.

- limitation may be needed (5 or so?)
- not as viewable as proposal 1
- may be easier to implement
- will use more unused space and not "push" the content down

![feature](https://user-images.githubusercontent.com/6490538/74918362-40dd5380-53c9-11ea-8afd-bdb9f762c48b.png)

### 3. Pinned List

The last option may be the simplest way. Pinning a repository just means, that it will stay on top of the repository list.

- very limited ui development
- context "this is a repo" is clearly visible
- is limited to the amount of "shown" pages
- defining a limit may be need (5 seems ok)

![list](https://user-images.githubusercontent.com/6490538/74918512-7da94a80-53c9-11ea-97d9-0051c17993c6.png)

**Comment by realpixelcode** [comment]
This would be really nice!

**Comment by lhinderberger** [comment]
This would also be very useful for individual user profiles, in addition to organizations.

**Comment by Mikaela** [comment]
I stumbled upon this with individual user profile as I have "real" repositories and then now I have a tutorial repository for learning. The result is that whenever I push to the tutorial repository it gets pushed to top and takes visibility from the real repositories that I would actually like to show.

**Comment by acoolstraw** [comment]
yo why haven't you implemented this yet

**Comment by era** [comment]
I may work on this feature if no one started yet, at least on the backend.

Regarding the backend, independent of the option we go with, I imagine we will have to:

- Add a `Pinned`flag to the repo table.
- Add a configurable* limit for the number of pinned repo

*not sure what this would mean, but I imagine different gitea owners may want different configurations for this, right?

Regarding the API, should we create a new API to pin? Or should we use the editRepo? Should users be able to pin while creating a new repo?

**Comment by acoolstraw** [comment]
I think pinning repos is more of a user thing, because multiple users should be able to pin a repo (if they've contributed to it). But I don't think they should be able to pin it when creating it

**Comment by Eekle** [comment]
I think I'd like to tackle this issue.

Does it satisfy the requirement if pinned repositories are simply listed first on a user/org page?

Do we actually need a limit on how many repositories can be pinned? If it's just a sorting feature, then it can be up to orgs/users to determine what suits them.

For the API, are we happy to just extend `PATCH /repos/{owner}/{repo}`?

Is it preferable to have `POST /repos/{owner}/{repo}/pin`, `DELETE /repos/{owner}/{repo}/pin`.

**Comment by lunny** [comment]
Maybe max limit is 6 and 3 for one line.

**Comment by Eekle** [comment]
Is this presuming that we use the "3 tiles" format, suggestion #1? I was going to go with #3 for ease of implementation.

**Comment by delvh** [comment]
> The only thing I can somewhat guess would be that you need a dropdown when pinning to decide where to pin it to, with the options being **you**, or the orgs you have admin rights in. If you don't have admin rights in any org, we should probably shortcut to directly pinning to your own page.

_Originally posted by @delvh in https://github.com/go-gitea/gitea/issues/19831#issuecomment-1293149783_

> It seems like you're suggesting that orgs should also be able to pin any repository to their profiles - not just ones tha they own?
I thought we were suggesting users could pin anything, but orgs could only pin their own repos.

_Originally posted by @Eekle in https://github.com/go-gitea/gitea/issues/19831#issuecomment-1293162834_

> > but orgs could only pin their own repos.

> Would that be *important* to limit though? Maybe it should also be unrestricted.

_Originally posted by @ell1e in https://github.com/go-gitea/gitea/issues/19831#issuecomment-1293171848_

I'm moving the discussion now here as it is more appropriate here.

So, at the moment, we have the following options:

- allow any owner to pin any repo they can see
- allow users to pin any repo they can see, orgs only their own repos
- allow every owner only to pin their own repos

**Comment by Eekle** [comment]
_Repeating my concern from https://github.com/go-gitea/gitea/pull/19831#issuecomment-1293180032_

I think the UI could really get in the way of function here. Pinning a repository to its owner's profile is going to be the most common use case, probably overwhelmingly common in the case of orgs.

If we allow arbitrary pinning to org+user profiles, I think it could easily muddy the experience of the common case.

---

Perhaps if the main pin button were for "pin/unpin from owner", but we include a dropdown for more intricate options (pin/unpin from active user, or from other administrated orgs), then that might be nice.

**Comment by delvh** [comment]
Yes, that was also what I meant.

---

Also, when we implement 1 or 2, we always have to filter all pinned repos for whether the current viewer can see the corresponding repos as we can otherwise have security issues where unallowed users can see that a specific repo exists.
Furthermore, the current approach of a denormalized data structure is not a good idea in that case as we probably need to use nested`JOIN`s if we don't want serious performance drawbacks.

**Comment by Eekle** [comment]
I suppose we have to check visibility no matter what, though? For things like private repos?

**Comment by delvh** [comment]
Makes sense. As I mentioned, the approach of storing a JSON list is in that case probably not a good idea anymore.
Unless of course you have an easy idea how to ensure that a user can really see a pinned repo with the current mechanism…

**Comment by Eekle** [comment]
No I agree with you. It feels like a `Pins` table may be in order.

**Comment by Eekle** [comment]
Smart. One thing to consider is that you might lose visibility of a repo you've pinned. For example if it gets made private, or you lose access to an org.

In those cases, I guess we can just leave the entries in the `PinnedRepo` table - and then the next time pins get updated for that user/org, we can clean up any invalid entries. That'll save the work of trying to work out all the operations that might invalidate pins.

**Comment by ell1e** [comment]
>  or you lose access to an org

Shouldn't a pin depend on 100% unrestricted public visibility anyway? If I have access to an org but nobody looking at my profile does, I'm not sure it should be allowed that I pin anything from that org. After all almost nobody can then go and look at it anyway, that would be very confusing.

**Comment by Eekle** [comment]
> > or you lose access to an org
> 
> Shouldn't a pin depend on 100% unrestricted public visibility anyway? If I have access to an org but nobody looking at my profile does, I'm not sure it should be allowed that I pin anything from that org. After all almost nobody can then go and look at it anyway, that would be very confusing.

It would be intuitive to me if you could pin whatever, but other users wouldn't see it on your profile unless they had visibility of it too.

**Comment by delvh** [comment]
Yep, that's also the approach I had in mind.

**Comment by ell1e** [comment]
Right, I guess that has some use especially in corporate instances where the majority might be private but still be a visible to a lot of people (and hence worth pinning). I can however see it leading to some confusing on mostly public instances of users pinning something that they thought was private, then wondering why that even worked and maybe not realizing it's not *really* shown on their profile anyway.

**Comment by delvh** [comment]
What is GitHubs approach here? Do they allow pinning private repos?

**Comment by delvh** [comment]
Also, pinning private repos makes even sense:
If you want to quickly find a repo again, that is used widely, but only internally in your organization, then pinning a private repo would be the equivalent of a bookmark/ quick access.

**Comment by Eekle** [comment]
> Right, I guess that has some use especially in corporate instances where the majority might be private but still be visible to a lot of people and hence worth pinning. I can however see it leading to some confusing on mostly public instances when users pin something there that they thought was (and is) private, then wondering why that even worked and maybe not realizing it's not _really_ shown on their profile anyway.

The current implementation marks a repo as `Private` in the pin card.

![image](https://user-images.githubusercontent.com/96976531/198252027-0fc78867-111d-42b5-a65e-6eb8fd772004.png)

This is the same marking used on the actual repository listing on the profile.

So I hope that is enough of a cue to a confused user that "Oh this is listed as private, so they've taken it into account."

**Comment by Eekle** [comment]
I suspect given the scope of changes suggested relative to the first implementation, it's going to have to be pretty much all torn out. The only reusable work is likely to be the pin card implementations on profile pages.

**Comment by ell1e** [comment]
> The current implementation marks a repo as Private in the pin card. [...] So I hope that is enough of a cue

Omg yes, that looks perfect! I love that. This is going to be pretty cool if it all works out as planned.

**Comment by BRAVO68WEB** [comment]
Hey guys, is their any update on this?

**Comment by Ashvith10** [comment]
What's the status of this issue?

**Comment by Ashvith10** [comment]
Haven't got any response yet, but I am still looking for a workaround. I have a few projects that I would like to highlight in my developer portfolio web-app. Now, this would take advantage of the Gitea/Codeberg API, where I'll be able to choose projects that will be shown on my website. In GitHub, you can do this conveniently by pinning and using that specific API to highlight pinned projects, but I'm not sure if there's any similar mechanisms for Gitea/Codeberg?

**Comment by ell1e** [comment]
@Ashvith10 there actually is kind of a workaround now! You can do both [profile readmes and org readmes](https://docs.gitea.com/next/usage/profile-readme) in the latest gitea version, which looks like this in practice: [(me on codeberg with profile readme)](https://codeberg.org/ell1e). and you can just put projects there! (the org readme update might not have reached codeberg yet, but it should soon.) see

**Comment by KaKi87** [comment]
> in the latest gitea version

Since v1.20.1, actually.

**Comment by ell1e** [comment]
Not the org readmes. Anyway, it should reach codeberg soon.

**Comment by Ashvith10** [comment]
@ell1e unfortunately, that isn't what I'm looking for, I was talking about the Codeberg REST API, actually. So, if there was a way to pin projects, I could access that particular API from the web app that I'm hosting on a cloud platform.

So far, I've come across this:

```console
curl -s https://codeberg.org/api/v1/user/starred -H "Authorization: token <hidden-for-security-reasons>"
```

But this also includes other starred users, which means pointless overheads.

**Comment by ell1e** [comment]
you can put into the HTML comments into a markdown file like the profile readme, and i think they render into the final output, usually. so it's possible to hide machine readable data in there.

**Comment by carlosfelgueiras** [comment]
Me and @DanielMatiasCarvalho are working on this.

**Comment by mrmelon54** [comment]
> So I hope that is enough of a cue to a confused user that "Oh this is listed as private, so they've taken it into account."

Do private repos only show up if the current user has view access to that repo?
